packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+import { assertNever } from "./lib/assert";
+import type { Observable } from "./lib/EventSource";
+import { makeEventSource } from "./lib/EventSource";
+import * as console from "./lib/fancy-console";
+import type { BuiltinEvent, Target } from "./lib/fsm";
+import { FSM } from "./lib/fsm";
+import type {
+  IWebSocketCloseEvent,
+  IWebSocketEvent,
+  IWebSocketInstance,
+  IWebSocketMessageEvent,
+} from "./types/IWebSocket";
+
+// TODO DRY this type up with the ConnectionStatus type in room.ts
+export type PublicConnectionStatus =
+  | "closed" // Room hasn't been entered, or has left already
+  | "authenticating" // Authentication has started, but not finished yet
+  | "connecting" // Authentication succeeded, now attempting to connect to a room
+  | "open" // Successful room connection, on the happy path
+  | "unavailable" // Connection lost unexpectedly, considered a temporary hiccup, will retry
+  | "failed"; // Connection failed and we won't retry automatically (e.g. unauthorized)
+
+/**
+ * Maps internal machine state to the public connection status API.
+ */
+function toPublicConnectionStatus(state: State): PublicConnectionStatus {
+  switch (state) {
+    case "@ok.connected":
+    case "@ok.awaiting-pong":
+      return "open";
+
+    case "@idle.initial":
+      return "closed";
+
+    case "@auth.busy":
+    case "@auth.backoff":
+      return "authenticating";
+
+    case "@connecting.busy":
+      return "connecting";
+
+    case "@connecting.backoff":
+      return "unavailable";
+
+    case "@idle.failed":
+      return "failed";
+
+    default:
+      return assertNever(state, "Unknown state");
+  }
+}
+
+/**
+ * Events that can be sent to the machine externally.
+ */
+type Event =
+  // Public events that can be called on the connection manager
+  | { type: "CONNECT" } // e.g. when trying to enter a room
+  | { type: "RECONNECT" } // e.g. user asking for an explicit reconnect of the socket
+  | { type: "DISCONNECT" } // e.g. leaving the room
+  | { type: "WINDOW_GOT_FOCUS" } // e.g. user's browser tab is refocused
+  | { type: "NAVIGATOR_ONLINE" } // e.g. browser gets back online
+
+  // Events that the connection manager will internally deal with
+  | { type: "PONG" }
+  | { type: "EXPLICIT_SOCKET_ERROR"; event: IWebSocketEvent }
+  | { type: "EXPLICIT_SOCKET_CLOSE"; event: IWebSocketCloseEvent }
+
+  // Only used by the E2E testing app, to simulate a pong timeout :(
+  | { type: "PONG_TIMEOUT" };
+
+type State =
+  | "@idle.initial"
+  | "@idle.failed"
+  | "@auth.busy"
+  | "@auth.backoff"
+  | "@connecting.busy"
+  | "@connecting.backoff"
+  | "@ok.connected"
+  | "@ok.awaiting-pong";
```

Comment:
FYI, there is no meaning to these `@`'s in the strings. I just found the convention useful to instantly tell state names apart from `EVENT_NAMES` and other string literals.

Reply:
I don't have a great reason for it, but I find myself wanting this to be an enum. Then you wouldn't really need to worry about differentiating with the `@` because you'd just have like `State.Idle.Initial` or something and you never really worry about the underlying string.

Reply:
Yeah, like you said in the other comment below, not sure how to combine that with the wildcard patterns. Those patterns are not just syntactic luxury, but an important construct that is needed for the machine.